### PR TITLE
Create a History of the changes inside a Template Layout/ Use this Hi…

### LINKED ISSUE #29148: Modifying the Template breaks pages on the receiver
### Parent Issue

_No response_

### Problem Statement

If you modify a template and PP it, the multi_tree table on the receiver won't be updated so the page will defer from the sender (or if the template originally had timestamp UUIDs, the page will be blank).

### Steps to Reproduce

1- Create a design template like this one:
Screenshot 2024-07-05 at 1 58 14 PM
3- Create a page with that template
4- Add a few contents
5- PP the created page = Pages look the same on sender and receiver.
Screenshot 2024-07-05 at 2 11 02 PM
6- Modify the template: Add a new box before the second row and should use the same container that will be on the second row.
Screenshot 2024-07-05 at 2 14 36 PM
7- PP the template using the filter Everything and Dependencies.
8- Compare the pages; the one of the senders looks different from the one of the receiver.

**Sender**
Screenshot 2024-07-05 at 2 18 14 PM

**Receiver**
Screenshot 2024-07-05 at 2 17 53 PM

**NOTE:**
The customer was coming from pretty old versions when template-containers UUIDs weren't sequential; when updating the template, it updated the UUIDs to sequential, so the page in the receiver shows blank due to the relation_type in the multi_tree table still using the old ones.

### Acceptance Criteria

The page should look the same on the sender and on the receiver.

When PP a template, if there were updates in the design, all pages that uses that template should be PP.

### dotCMS Version

Tested on 24.04.24 LTS and demo, but probably all LTS are affected.

### Proposed Objective

Customer Success

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

https://dotcms.zendesk.com/agent/tickets/116560

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

The current workaround is that when PP a template, Force push all impacted pages.

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by freddyDOTCMS**
The problem here is that when you send a TemplateLayout after changed it to a receiver we don't have a way to know witch was the movement on the receiver because the UUID are already all recalculated, and the multi_tree are not sent unless you sent the page too.

The fix was create a new history on the layout to know how the UUID of each Container have been move, also I created a new version attribute to count how many time the Layout have changed.

### Proposed Changes
* Change in the Handler to use the method to update the Layout when the Template already exists on the receiver

https://github.com/dotCMS/core/pull/29752/files#diff-1e5a19fcdfdbe883752308cb7adf42231e860dff8300edf8b20c66f8b2b0fd39R168-R185

This method also recalculated the UUID.

* Create a new History field

https://github.com/dotCMS/core/pull/29752/files#diff-8a7533dd3f0219197b317cfcfc4848a3e61492c38f1bac1d5aeaec566025c626R50

* keep the history when update a TemplateLayout

https://github.com/dotCMS/core/pull/29752/files#diff-ce9702cf145c3dc3fd6472a323188b4897be3c98058272d8eba3f54d836f46ceR1433

* Create a new version field

https://github.com/dotCMS/core/pull/29752/files#diff-edc3e1981140c3dc84b3fbab35e5afb0ffc0e9f3f8b47673ed42da6939bf83fbR58

* recalculated the version when the TemplateLayout is saved

https://github.com/dotCMS/core/pull/29752/files#diff-ce9702cf145c3dc3fd6472a323188b4897be3c98058272d8eba3f54d836f46ceR1306

* Calculate the changes using the new History attribute on the receiver

https://github.com/dotCMS/core/pull/29752/files#diff-ce9702cf145c3dc3fd6472a323188b4897be3c98058272d8eba3f54d836f46ceR1319

https://github.com/dotCMS/core/pull/29752/files#diff-acae9e18f11a2e8737fad0e31d65f09f2e330c049e6320996de89ed5ef495ef4R26

### Checklist
- [ ] Tests
- [ ] Translations
- [ ] Security Implications Contemplated (add notes if applicable)

### Additional Info
** any additional useful context or info **

### Screenshots
Original             |  Updated
:-------------------------:|:-------------------------:
** original screenshot **  |  ** updated screenshot **

This PR fixes: #29148

### CODE REVIEW

**Review by jcastro-dotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/portlets/templates/design/bean/ContainerUUID.java [hunk 1]
```diff
@@ -23,11 +30,58 @@ public class ContainerUUID implements Serializable{
...
+                         final @JsonProperty("historyUUIDs") List<String> historyUUIDs) {
 
         this.identifier = containerIdOrPath;
         this.uuid = containerInstanceID == null ? UUID_DEFAULT_VALUE : containerInstanceID;
+
+        if (isNew(containerInstanceID)) {
+            this.historyUUIDs = new ArrayList<>();
+        } else {
+            this.historyUUIDs = UtilMethods.isSet(historyUUIDs) ? new ArrayList<>(historyUUIDs) : list(this.uuid);
+            ;
```
Comment (victoralfaro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/templates/design/bean/ContainerUUID.java:71]: Remove this colon.

File: dotcms-integration/src/test/java/com/dotcms/datagen/TemplateLayoutDataGen.java [hunk 2]
```diff
@@ -5,18 +5,20 @@
...
 public class TemplateLayoutDataGen  {
 
     private List<TemplateLayoutRow> rows;
 
     private List<TemplateLayoutColumn> currentColumns;
 
-    private Map<String, List<String>> containersIds = new HashMap<>();
-    final Map<String, List<String>> containersIdsInSidebar = new HashMap<>();
+    private List<ContainerUUID> containersIds = new ArrayList<>();
+    final List<ContainerUUID> containersIdsInSidebar = new ArrayList<>();
```
Comment (victoralfaro-dotcms) [inline dotcms-integration/src/test/java/com/dotcms/datagen/TemplateLayoutDataGen.java:19]: This can be private.

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/portlets/templates/design/bean/ContainerUUID.java (hunk 1 above)
Comment (freddyDOTCMS) [reply dotCMS/src/main/java/com/dotmarketing/portlets/templates/design/bean/ContainerUUID.java]: done https://github.com/dotCMS/core/pull/29752/commits/b16f6562e59d264efd3c344bd7164dec250da377

File: dotcms-integration/src/test/java/com/dotcms/datagen/TemplateLayoutDataGen.java (hunk 2 above)
Comment (freddyDOTCMS) [reply dotcms-integration/src/test/java/com/dotcms/datagen/TemplateLayoutDataGen.java]: done https://github.com/dotCMS/core/pull/29752/commits/b5a1836872a6ccdd83eac1ef97876df83d0f56d8

**Review by victoralfaro-dotcms — APPROVED** [review]

**Pull request merged**